### DISCUSSION TITLE
An example or conversion on how to use collab with signalR ?

### DISCUSSION THREAD

**Discussion by voltcode** [stepwisehq/prosemirror-dotnet#12] [General]
I'm wondering whether the code in this repo can be used to implement  something akin to:
https://github.com/ProseMirror/website/tree/master/src/collab

with SignalR/WS instead of http calls and polling?

Do you have any pointers on how to approach this issue? Which npm module should I use in JS? Will the original work or do I need to use the commit one?

**Comment by ProTip** [comment]
Hi @voltcode ,

Yes this package can be used in the creation of a collaborative editing backend as we have done for the [StepWise product](https://stepwisehq.com).

If following the ProseMirror collab example closely, the DTOs in this package can be used for serialization. For more complicated systems it can be use for validating updates, applying updates server-side, and etc as per the `README` examples.

Unfortunately we don't have a comprehensive example at this time and likely wouldn't until such a time as the StepWise source being made available.

I've created a number of ProseMirror backends and they have all been highly custom. Best bet is to follow the ProseMirror collab example closely as a starter and add logic to store/retrieve the updates in a database. After that you could look to existing OSS projects for inspiration on creating more featureful backends with versioning, snapshots, and etc.

**Comment by voltcode** [comment]
Ok thanks for the pointers. 
What I'm struggling a bit with is whether the server really has to apply the steps ? I understand it needs to store the full doc, because new collaborators may come in, but I could as well send entire doc from time to time from one client to the server (or replay it from edits) - it is far from optimal but for a PoC should do.

Are you aware of any other reasons why would the server have to apply the steps too ?

**Comment by ProTip** [comment]
Hi @voltcode ,

The use cases are probably pretty advanced and geared more towards productionizing a mature backend feature set. I wrote a bit about this in the announcement article [Announcing ProseMirror.Net](https://stepwisehq.com/blog/2023-07-17-announcing-prosemirror-net/) .

Here are some reasons you may want to use this project:

* Validation: With backend validation we can assure that the steps don’t violate the schema. Bugs in implementation on the frontend can cause invalid steps to be submitted to the backend
* Creating or modifying documents server-side
* Snapshotting documents for performance and versioning
* Server-side step mapping
